File: src/diffler/core/engine.py
"""Main orchestration engine."""

from __future__ import annotations

import logging
from pathlib import Path

from diffler.config import DifflerConfig
from diffler.core.context import ContextBuilder
from diffler.core.git import README_PATH, commit_and_push, has_changes
from diffler.core.renderer import Renderer

logger = logging.getLogger(__name__)


class Engine:
    """Orchestrates loading, rendering, validation, and committing."""

    def __init__(self, config: DifflerConfig, renderer: Renderer) -> None:
        self.config = config
        self.renderer = renderer
        self.context_builder = ContextBuilder(config)

    def render(self) -> str:
        """Render the profile README."""
        template_source = self.renderer.read_template_source()
        context = self.context_builder.build(template_source)
        return self.renderer.render(context)

    def validate(self) -> None:
        """Validate templates and configuration."""
        template_source = self.renderer.read_template_source()
        context = self.context_builder.build(template_source)
        self.renderer.validate(context)
        # Future: validate config schema, check GitHub token scopes, etc.

    def update(
        self,
        *,
        dry_run: bool = False,
        message: str = "🤖 Auto-update profile README",
    ) -> None:
        """Render and optionally commit the profile README."""
        rendered = self.render()

        if dry_run:
            logger.info("Dry run — output:\n%s", rendered)
            return

        current = ""
        if README_PATH.exists():
            current = README_PATH.read_text(encoding="utf-8")

        if rendered == current:
            logger.info("No changes detected; skipping commit.")
            return

        README_PATH.write_text(rendered, encoding="utf-8")
        logger.info("README.md updated locally.")

        if has_changes(README_PATH):
            commit_and_push(message, README_PATH)
        else:
            logger.info("No git changes detected; skipping commit.")

File: packages/diffler/src/core/engine.ts
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import type { DifflerConfig } from "../config.js";
import { commitAndPush, hasChanges, README_PATH } from "./git.js";
import { Renderer } from "./renderer.js";
import { ContextBuilder } from "./context.js";

export class Engine {
  private renderer: Renderer;
  contextBuilder: ContextBuilder;

  constructor(config: DifflerConfig, renderer: Renderer) {
    this.renderer = renderer;
    this.contextBuilder = new ContextBuilder(config);
  }

  async render(): Promise<string> {
    const templateSource = this.renderer.readTemplateSource();
    const context = await this.contextBuilder.build(templateSource);
    return this.renderer.render(context);
  }

  async validate(): Promise<void> {
    const templateSource = this.renderer.readTemplateSource();
    const context = await this.contextBuilder.build(templateSource);
    this.renderer.validate(context);
  }

  async update(options: { dryRun?: boolean; message?: string } = {}): Promise<void> {
    const rendered = await this.render();

    if (options.dryRun) {
      console.info("Dry run — output:\n", rendered);
      return;
    }

    let current = "";
    if (existsSync(README_PATH)) {
      current = readFileSync(README_PATH, "utf-8");
    }

    if (rendered === current) {
      console.info("No changes detected; skipping commit.");
      return;
    }

    writeFileSync(README_PATH, rendered, "utf-8");
    console.info("README.md updated locally.");

    if (hasChanges(README_PATH)) {
      commitAndPush(options.message || "🤖 Auto-update profile README", README_PATH);
    } else {
      console.info("No git changes detected; skipping commit.");
    }
  }
}
